app/src/main/java/com/google/cloud/solutions/flexenv/common/SpeechTranslationHelper.java

```diff
@@ -0,0 +1,146 @@
+                                      int sampleRateInHertz,
+                                      SpeechTranslationListener translationListener) {
+        JSONObject requestBody = new JSONObject();
+        try {
+            requestBody.put("encoding", SPEECH_TRANSLATE_ENCODING);
+            requestBody.put("sampleRateHertz", sampleRateInHertz);
+            requestBody.put("languageCode", context.getResources().getConfiguration().getLocales().get(0));
+            requestBody.put("audioContent", base64EncodedAudioMessage);
+        } catch(JSONException e) {
+            Log.e(TAG, e.getLocalizedMessage());
+            translationListener.onTranslationFailed(e);
+        }
+
+        byte[] requestBodyBytes = requestBody.toString().getBytes();
+        UrlRequest request = buildSpeechTranslationRequest(context, requestBodyBytes, translationListener);
+        request.start();
+    }
+
+    private UrlRequest buildSpeechTranslationRequest(Context context, byte[] requestBody, SpeechTranslationListener translationListener) {
+        CronetEngine.Builder myBuilder = new CronetEngine.Builder(context);
+        CronetEngine cronetEngine = myBuilder.build();
+        Executor executor = Executors.newSingleThreadExecutor();
+        String speechTranslateEndpoint = context.getString(R.string.speechToSpeechEndpoint);
+        UrlRequest.Builder requestBuilder = cronetEngine.newUrlRequestBuilder(
+                speechTranslateEndpoint, new UrlRequest.Callback(){
+                    StringBuilder responseBodyBuilder;
+                    @Override
+                    public void onRedirectReceived(UrlRequest request, UrlResponseInfo info, String newLocationUrl) {
+                        Log.i(TAG, "onRedirectReceived method called.");
+                        request.followRedirect();
+                    }
+                    @Override
+                    public void onResponseStarted(UrlRequest request, UrlResponseInfo info) {
+                        Log.i(TAG, "onResponseStarted method called.");
+                        int httpStatusCode = info.getHttpStatusCode();
+                        if (httpStatusCode == 200 || httpStatusCode == 400) {
+                            responseBodyBuilder = new StringBuilder();
+                            request.read(ByteBuffer.allocateDirect((int)info.getReceivedByteCount()));
+                        } else  {
+                            String errorMessage = "Unexpected HTTP status code: " + httpStatusCode;
+                            translationListener.onTranslationFailed(new SpeechTranslationException(errorMessage));
+                        }
+                    }
+                    @Override
+                    public void onReadCompleted(UrlRequest request, UrlResponseInfo info, ByteBuffer byteBuffer) {
+                        Log.i(TAG, "onReadCompleted method called.");
+                        request.read(ByteBuffer.allocateDirect((int)info.getReceivedByteCount()));
```

Comment:
nit: rather than allocate a new buffer each time, you could
`￼                         byteBuffer.flip();`
`￼                         responseBodyBuilder.append(Charset.forName("UTF-8").decode(byteBuffer).toString());`
￼`                         byteBuffer.clear();`
`￼                         request.read(byteBuffer);`

Reply:
That makes sense. I'll try this next.

Reply:
Implemented in 99815d6b70755e7ad8493070b083f78b5f03486e. Tests passing okay. Thank you!